File: backend/app/models/monitoring.py
from pydantic import BaseModel
from typing import List, Optional
from enum import Enum


class StatusType(str, Enum):
    GREEN = "🟢"
    YELLOW = "🟡"
    RED = "🔴"
    GREY = "⚪"


class UrlMapMetric(BaseModel):
    project_id: str
    url_map_name: str
    hostname: str
    http_status: Optional[int]
    status: str
    error: Optional[str] = None


class PodMetric(BaseModel):
    project_id: str
    cluster_name: str
    namespace: str
    pod_name: str
    status: str
    status_icon: str


class PubSubMetric(BaseModel):
    project_id: str
    subscription_name: str
    unacked_messages: int
    oldest_message_age_minutes: float
    status: str


class NodePoolMetric(BaseModel):
    project_id: str
    cluster_name: str
    node_pool_name: str
    current_nodes: int
    max_nodes: int
    utilization_percent: float
    status: str
    is_regional: bool


class PodRestartMetric(BaseModel):
    project_id: str
    cluster_name: str
    namespace: str
    pod_name: str
    restart_count: int
    status: str


class LatencyMetric(BaseModel):
    project_id: str
    backend_service: str
    p95_latency_seconds: float
    status: str


class SpannerMetric(BaseModel):
    project_id: str
    instance_name: str
    metric_type: str
    value_percent: float
    status: str


class MonitoringResponse(BaseModel):
    url_maps: List[UrlMapMetric] = []
    pods: List[PodMetric] = []
    pubsub: List[PubSubMetric] = []
    node_pools: List[NodePoolMetric] = []
    pod_restarts: List[PodRestartMetric] = []
    latency: List[LatencyMetric] = []
    spanner: List[SpannerMetric] = []
    timestamp: str
    errors: List[str] = []

File: frontend/src/types/monitoring.ts
export interface UrlMapMetric {
  project_id: string;
  url_map_name: string;
  hostname: string;
  http_status: number | null;
  status: string;
  error?: string;
}

export interface PodMetric {
  project_id: string;
  cluster_name: string;
  namespace: string;
  pod_name: string;
  status: string;
  status_icon: string;
}

export interface PubSubMetric {
  project_id: string;
  subscription_name: string;
  unacked_messages: number;
  oldest_message_age_minutes: number;
  status: string;
}

export interface NodePoolMetric {
  project_id: string;
  cluster_name: string;
  node_pool_name: string;
  current_nodes: number;
  max_nodes: number;
  utilization_percent: number;
  status: string;
  is_regional: boolean;
}

export interface PodRestartMetric {
  project_id: string;
  cluster_name: string;
  namespace: string;
  pod_name: string;
  restart_count: number;
  status: string;
}

export interface LatencyMetric {
  project_id: string;
  backend_service: string;
  p95_latency_seconds: number;
  status: string;
}

export interface SpannerMetric {
  project_id: string;
  instance_name: string;
  metric_type: string;
  value_percent: number;
  status: string;
}

export interface MonitoringResponse {
  url_maps: UrlMapMetric[];
  pods: PodMetric[];
  pubsub: PubSubMetric[];
  node_pools: NodePoolMetric[];
  pod_restarts: PodRestartMetric[];
  latency: LatencyMetric[];
  spanner: SpannerMetric[];
  timestamp: string;
  errors: string[];
}
